File: lib/nodets/test/test_handler.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * 'License'); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

//This is the server side Node test handler for the standard
//  Apache Thrift test service.

import ttypes = require("./gen-nodejs/ThriftTest_types");
import thrift = require("thrift");
import Thrift = thrift.Thrift;
import Int64 = require("node-int64");
import { v4 as uuid } from "uuid";
type uuid = string;

export class SyncThriftTestHandler {
  testVoid(): Promise<void> {
    //console.log('testVoid()');
    return Promise.resolve();
  }
  testMapMap(hello: number) {
    //console.log('testMapMap(' + hello + ')');

    var mapmap: { [key: number]: { [key: number]: number } } = [];
    var pos: { [key: number]: number } = [];
    var neg: { [key: number]: number } = [];
    for (var i = 1; i < 5; i++) {
      pos[i] = i;
      neg[-i] = -i;
    }
    mapmap[4] = pos;
    mapmap[-4] = neg;

    return Promise.resolve(mapmap);
  }
  testInsanity(argument: ttypes.Insanity): Promise<{ [k: number]: any }> {
    const first_map: { [k: number]: any } = [];
    const second_map: { [k: number]: any } = [];

    first_map[ttypes.Numberz.TWO] = argument;
    first_map[ttypes.Numberz.THREE] = argument;

    const looney = new ttypes.Insanity();
    second_map[ttypes.Numberz.SIX] = looney;

    const insane: { [k: number]: any } = [];
    insane[1] = first_map;
    insane[2] = second_map;

    return Promise.resolve(insane);
  }
  testMulti(
    arg0: any,
    arg1: number,
    arg2: Int64,
    arg3: { [k: number]: string },
    arg4: ttypes.Numberz,
    arg5: number,
  ) {
    var hello = new ttypes.Xtruct();
    hello.string_thing = "Hello2";
    hello.byte_thing = arg0;
    hello.i32_thing = arg1;
    hello.i64_thing = arg2;
    return Promise.resolve(hello);
  }
  testException(arg: string): Promise<void> {
    if (arg === "Xception") {
      var x = new ttypes.Xception();
      x.errorCode = 1001;
      x.message = arg;
      throw x;
    } else if (arg === "TException") {
      throw new Thrift.TException(arg);
    } else {
      return Promise.resolve();
    }
  }
  testMultiException(arg0: string, arg1: string) {
    if (arg0 === "Xception") {
      var x = new ttypes.Xception();
      x.errorCode = 1001;
      x.message = "This is an Xception";
      throw x;
    } else if (arg0 === "Xception2") {
      var x2 = new ttypes.Xception2();
      x2.errorCode = 2002;
      x2.struct_thing = new ttypes.Xtruct();
      x2.struct_thing.string_thing = "This is an Xception2";
      throw x2;
    }

    var res = new ttypes.Xtruct();
    res.string_thing = arg1;
    return Promise.resolve(res);
  }
  testOneway(sleepFor: number) {}

  testString(thing: string) {
    return Promise.resolve(thing);
  }
  testBool(thing: boolean) {
    return Promise.resolve(thing);
  }
  testByte(thing: number) {
    return Promise.resolve(thing);
  }
  testI32(thing: number) {
    return Promise.resolve(thing);
  }
  testI64(thing: number) {
    return Promise.resolve(thing);
  }
  testDouble(thing: number) {
    return Promise.resolve(thing);
  }
  testBinary(thing: Buffer) {
    return Promise.resolve(thing);
  }
  testUuid(thing: uuid) {
    return Promise.resolve(thing);
  }
  testStruct(thing: ttypes.Xtruct) {
    return Promise.resolve(thing);
  }
  testNest(thing: ttypes.Xtruct2) {
    return Promise.resolve(thing);
  }
  testMap(thing: { [k: number]: number }) {
    return Promise.resolve(thing);
  }
  testStringMap(thing: { [k: string]: string }) {
    return Promise.resolve(thing);
  }
  testSet(thing: number[]) {
    return Promise.resolve(thing);
  }
  testList(thing: number[]) {
    return Promise.resolve(thing);
  }
  testEnum(thing: ttypes.Numberz) {
    return Promise.resolve(thing);
  }
  testTypedef(thing: number) {
    return Promise.resolve(thing);
  }
}

export class AsyncThriftTestHandler {
  private syncHandler: SyncThriftTestHandler;
  constructor() {
    this.syncHandler = new SyncThriftTestHandler();
  }

  testVoid(callback: (result: void) => void): Promise<void> {
    callback(undefined);
    return Promise.resolve();
  }
  testMapMap(
    hello: number,
    callback: (
      err: any,
      result: { [k: number]: { [k: number]: number } },
    ) => void,
  ): Promise<void> {
    var mapmap: { [key: number]: { [key: number]: number } } = [];
    var pos: { [key: number]: number } = [];
    var neg: { [key: number]: number } = [];
    for (var i = 1; i < 5; i++) {
      pos[i] = i;
      neg[-i] = -i;
    }
    mapmap[4] = pos;
    mapmap[-4] = neg;

    callback(null, mapmap);
    return Promise.resolve();
  }
  testInsanity(
    argument: ttypes.Insanity,
    callback?: (err: any, result: { [k: number]: any }) => void,
  ): Promise<void> {
    const first_map: { [k: number]: any } = [];
    const second_map: { [k: number]: any } = [];

    first_map[ttypes.Numberz.TWO] = argument;
    first_map[ttypes.Numberz.THREE] = argument;

    const looney = new ttypes.Insanity();
    second_map[ttypes.Numberz.SIX] = looney;

    const insane: { [k: number]: any } = [];
    insane[1] = first_map;
    insane[2] = second_map;

    if (callback !== undefined) {
      callback(null, insane);
    }
    return Promise.resolve();
  }
  testMulti(
    arg0: any,
    arg1: number,
    arg2: Int64,
    arg3: { [k: number]: string },
    arg4: ttypes.Numberz,
    arg5: number,
    result: Function,
  ): Promise<void> {
    var hello = this.syncHandler.testMulti(arg0, arg1, arg2, arg3, arg4, arg5);
    hello.then((hello) => result(null, hello));
    return Promise.resolve();
  }
  testException(arg: string, result: (err: any) => void): Promise<void> {
    if (arg === "Xception") {
      var x = new ttypes.Xception();
      x.errorCode = 1001;
      x.message = arg;
      result(x);
    } else if (arg === "TException") {
      result(new Thrift.TException(arg));
    } else {
      result(null);
    }
    return Promise.resolve();
  }
  testMultiException(
    arg0: string,
    arg1: string,
    result: (err: any, res?: ttypes.Xtruct) => void,
  ): Promise<void> {
    if (arg0 === "Xception") {
      var x = new ttypes.Xception();
      x.errorCode = 1001;
      x.message = "This is an Xception";
      result(x);
    } else if (arg0 === "Xception2") {
      var x2 = new ttypes.Xception2();
      x2.errorCode = 2002;
      x2.struct_thing = new ttypes.Xtruct();
      x2.struct_thing.string_thing = "This is an Xception2";
      result(x2);
    } else {
      var res = new ttypes.Xtruct();
      res.string_thing = arg1;
      result(null, res);
    }
    return Promise.resolve();
  }
  testOneway(sleepFor: number, result: Function) {
    this.syncHandler.testOneway(sleepFor);
  }
  testString(
    thing: string,
    callback: (err: any, result: string) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testByte(
    thing: number,
    callback: (err: any, result: number) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testBool(
    thing: boolean,
    callback: (err: any, result: boolean) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testI32(
    thing: number,
    callback: (err: any, result: number) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testI64(
    thing: number,
    callback: (err: any, result: number) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testDouble(
    thing: number,
    callback: (err: any, result: number) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testBinary(
    thing: Buffer,
    callback: (err: any, result: Buffer) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testUuid(
    thing: uuid,
    callback: (err: any, result: uuid) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testStruct(
    thing: ttypes.Xtruct,
    callback: (err: any, result: ttypes.Xtruct) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testNest(
    thing: ttypes.Xtruct2,
    callback: (err: any, result: ttypes.Xtruct2) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testMap(
    thing: { [k: number]: number },
    callback: (err: any, result: { [k: number]: number }) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testStringMap(
    thing: { [k: string]: string },
    callback: (err: any, result: { [k: string]: string }) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testSet(
    thing: number[],
    callback: (err: any, result: number[]) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testList(
    thing: number[],
    callback: (err: any, result: number[]) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testEnum(
    thing: ttypes.Numberz,
    callback: (err: any, result: ttypes.Numberz) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
  testTypedef(
    thing: number,
    callback: (err: any, result: number) => void,
  ): Promise<void> {
    callback(null, thing);
    return Promise.resolve();
  }
}

File: lib/kotlin/cross-test-server/src/main/kotlin/org/apache/thrift/test/TestHandler.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.thrift.test

import java.nio.ByteBuffer
import kotlinx.coroutines.delay
import org.apache.thrift.TException
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import thrift.test.Insanity
import thrift.test.Numberz
import thrift.test.ThriftTest
import thrift.test.Xception
import thrift.test.Xception2
import thrift.test.Xtruct
import thrift.test.Xtruct2

class TestHandler : ThriftTest {

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(TestHandler::class.java)
    }

    override suspend fun testVoid() {
        logger.info("testVoid()\n")
    }

    override suspend fun testString(thing: String): String {
        logger.info("testString(\"$thing\")\n")
        return thing
    }

    override suspend fun testBool(thing: Boolean): Boolean {
        logger.info("testBool($thing)\n")
        return thing
    }

    override suspend fun testByte(thing: Byte): Byte {
        logger.info("testByte($thing)\n")
        return thing
    }

    override suspend fun testI32(thing: Int): Int {
        logger.info("testI32($thing)\n")
        return thing
    }

    override suspend fun testI64(thing: Long): Long {
        logger.info("testI64($thing)\n")
        return thing
    }

    override suspend fun testDouble(thing: Double): Double {
        logger.info("testDouble($thing)\n")
        return thing
    }

    override suspend fun testBinary(thing: ByteArray): ByteArray {
        val buffer = ByteBuffer.wrap(thing)
        val sb = StringBuilder(buffer.remaining() * 3)
        buffer.mark()
        var limit = 0 // limit output to keep the log size sane
        while (buffer.remaining() > 0 && ++limit < 1024) {
            sb.append(String.format("%02X ", buffer.get()))
        }
        if (buffer.remaining() > 0) {
            sb.append("...") // indicate we have more date
        }
        logger.info("testBinary($sb)\n")
        buffer.reset()
        return buffer.array()
    }

    override suspend fun testStruct(thing: Xtruct): Xtruct {
        logger.info(
            """
testStruct({"${thing.string_thing}", ${thing.byte_thing}, ${thing.i32_thing}, ${thing.i64_thing}})

""".trimIndent()
        )
        return thing
    }

    override suspend fun testNest(thing: Xtruct2): Xtruct2 {
        val thing2: Xtruct = thing.struct_thing!!
        logger.info(
            """
testNest({${thing.byte_thing}, {"${thing2.string_thing}", ${thing2.byte_thing}, ${thing2.i32_thing}, ${thing2.i64_thing}}, ${thing.i32_thing}})

""".trimIndent()
        )
        return thing
    }

    override suspend fun testMap(thing: Map<Int, Int>): Map<Int, Int> {
        logger.info("testMap({")
        logger.info("{}", thing)
        logger.info("})\n")
        return thing
    }

    override suspend fun testStringMap(thing: Map<String, String>): Map<String, String> {
        logger.info("testStringMap({")
        logger.info("{}", thing)
        logger.info("})\n")
        return thing
    }

    override suspend fun testSet(thing: Set<Int>): Set<Int> {
        logger.info("testSet({")
        var first = true
        for (elem in thing) {
            if (first) {
                first = false
            } else {
                logger.info(", ")
            }
            logger.info("{}", elem)
        }
        logger.info("})\n")
        return thing
    }

    override suspend fun testList(thing: List<Int>): List<Int> {
        logger.info("testList({")
        var first = true
        for (elem in thing) {
            if (first) {
                first = false
            } else {
                logger.info(", ")
            }
            logger.info("{}", elem)
        }
        logger.info("})\n")
        return thing
    }

    override suspend fun testEnum(thing: Numberz): Numberz {
        logger.info("testEnum($thing)\n")
        return thing
    }

    override suspend fun testTypedef(thing: Long): Long {
        logger.info("testTypedef($thing)\n")
        return thing
    }

    override suspend fun testMapMap(hello: Int): Map<Int, Map<Int, Int>> {
        logger.info("testMapMap($hello)\n")
        val mapmap: MutableMap<Int, Map<Int, Int>> = HashMap()
        val pos = HashMap<Int, Int>()
        val neg = HashMap<Int, Int>()
        for (i in 1..4) {
            pos[i] = i
            neg[-i] = -i
        }
        mapmap[4] = pos
        mapmap[-4] = neg
        return mapmap
    }

    override suspend fun testInsanity(argument: Insanity): Map<Long, Map<Numberz, Insanity>> {
        logger.info("testInsanity()\n")
        val firstMap = mutableMapOf<Numberz, Insanity>()
        val secondMap = mutableMapOf<Numberz, Insanity>()
        firstMap[Numberz.TWO] = argument
        firstMap[Numberz.THREE] = argument
        val looney = Insanity()
        secondMap[Numberz.SIX] = looney
        val insane: MutableMap<Long, Map<Numberz, Insanity>> = HashMap()
        insane[1L] = firstMap
        insane[2L] = secondMap
        return insane
    }

    override suspend fun testMulti(
        arg0: Byte,
        arg1: Int,
        arg2: Long,
        arg3: Map<Short, String>,
        arg4: Numberz,
        arg5: Long
    ): Xtruct {
        logger.info("testMulti()\n")
        val hello = Xtruct()
        hello.string_thing = "Hello2"
        hello.byte_thing = arg0
        hello.i32_thing = arg1
        hello.i64_thing = arg2
        return hello
    }

    @Throws(Xception::class, TException::class)
    override suspend fun testException(arg: String) {
        logger.info("testException($arg)\n")
        when (arg) {
            "Xception" -> {
                val x = Xception()
                x.errorCode = 1001
                x.setFieldValue(Xception._Fields.MESSAGE, arg)
                throw x
            }
            "TException" -> {
                // Unspecified exception should yield a TApplicationException on client side
                throw RuntimeException(arg)
            }
            else -> {
                val result = Xtruct()
                result.string_thing = arg
            }
        }
        return
    }

    @Throws(Xception::class, Xception2::class)
    override suspend fun testMultiException(arg0: String, arg1: String): Xtruct {
        logger.info("testMultiException($arg0, $arg1)\n")
        if (arg0 == "Xception") {
            val x = Xception()
            x.errorCode = 1001
            x.setFieldValue(Xception._Fields.MESSAGE, "This is an Xception")
            throw x
        } else if (arg0 == "Xception2") {
            val x = Xception2()
            x.errorCode = 2002
            x.struct_thing = Xtruct().apply { string_thing = "This is an Xception2" }
            throw x
        }
        val result = Xtruct()
        result.string_thing = arg1
        return result
    }

    override suspend fun testOneway(secondsToSleep: Int) {
        logger.info("testOneway($secondsToSleep) => sleeping...")
        try {
            delay(secondsToSleep * 1000L)
            logger.info("Done sleeping!")
        } catch (ie: InterruptedException) {
            throw RuntimeException(ie)
        }
    }
}
